VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To check if a "tested version" is contained within a version range:
+
+- Start from a parsed a version range specifier with:
+
+  - a versioning scheme
+  - a list of constraints of comparator and version, sorted by version
+    and where each version occurs only once in any constraint.
+
+- If the constraint list contains only one item and the comparator is "*",
+  then the "tested version" is IN the range. Check is finished.
+
+- Select the version equality and comparison procedures suitable for this
+  versioning scheme and use these for all version comparisons performed below.
+
+- If the "tested version" is equal to the any of the constraint version
+  where the constraint comparator is for equality (any of "=", "<=", or ">=")
+  then the "tested version" is in the range. Check is finished.
+
+- If the "tested version" is equal to the any of the constraint version where
+  the constraint comparator is "=!" then the "tested version" is NOT in the
+  range. Check is finished.
+
+- Split the constraint list in two sub lists:
+
+  - a first list where the comparator is "=" or "!="
+  - a second list where the comparator is neither "=" nor "!="
+
+- Iterate over the current and next contiguous constraints pairs (aka. pairwise)
+  in the second list.
+
+- For each current and next constraint:
+
+    - If this is the first iteration and current comparator is "<" or <="
+      and the "tested version" is less than the current version
+      then the "tested version" is IN the range. Check is finished.
+
+    - If this is the last iteration and next comparator is ">" or >="
+      and the "tested version" is greater than the next version
+      then the "tested version" is IN the range. Check is finished.
+
+    - If current comparator is ">" or >=" and next comparator is "<" or <="
+      and the "tested version" is greater than the current version
+      and the "tested version" is less than the next version
+      then the "tested version" is IN the range. Check is finished.
+
+    - If current comparator is "<" or <=" and next comparator is ">" or >="
+      then these versions are out the range. Continue to the next iteration.
+
+- Reaching here without having finished the check before means that the
+  "tested version" is NOT in the range.
+
+
+Notes and caveats
+~~~~~~~~~~~~~~~~~~~
+
+- Comparing versions from two different versioning schemes is an error. Even
+  though there may be some similarities between the ``semver`` version of an npm
+  and the ``deb`` version of its Debian packaging, the way versions are compared
+  specific to each versioning scheme and may be different. Tools should report
+  an error in this case.
+
+- All references to sorting or ordering of version constraints means sorting
+  by version. And sorting by versions always implies using the versioning
+  scheme-specified version comparison and ordering.
+
+
+Some of the known versioning schemes
```

Comment:
Like the PURL spec, these schemes should be separated into their own document.

Based on experience with PURL, I think the spec should be written with the intention that not all implementations support all schemes, and implementations should not try to support schemes without understanding them unless this spec is accompanied by a comprehensive and well researched test suite. For PURL, there are a lot of cases that the test suite does not cover, and the test suite in at least one case asserts incorrect behavior because the ecosystem behavior wasn't well understood when it was added to the spec, and because the incorrect behavior is in the test suite it has been propagated into multiple implementations.

Reply:
> Like the PURL spec, these schemes should be separated into their own document.

Good points!

Reply:
We can do that soon after I merge this today.